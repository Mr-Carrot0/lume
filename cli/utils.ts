import lume from "../mod.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(_config?: URL): Promise<Site> {
  if (!_config) {
    return lume();
  }
  log.info(`Loading config file <gray>${_config}</gray>`);
  const mod = await import(_config.toString());
  const site = mod.default as Site | undefined;
  if (!(site instanceof Site)) {
    log.fatal(
      `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
    );
    throw new Error("Site instance is not found");
  }
  return site;
}

/** Create a site intance and build it */
export async function buildSite(site: Site): Promise<void> {
  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );
}
