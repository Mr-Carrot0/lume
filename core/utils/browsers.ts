/**
 * Code generated from:
 *
 * ```ts
 * import { getCompatibleVersions } from "npm:baseline-browser-mapping@2.4.4";
 *
 * const results = getCompatibleVersions({
 *   widelyAvailableOnDate: "2025-05-07", // Lume 3.0.0 release date
 * });
 * ```
 *
 * Not run directly to avoid importing this package that requires read access.
 */
export const browsers = {
  chrome: [107, 0],
  chrome_android: [107, 0],
  edge: [107, 0],
  firefox: [104, 0],
  firefox_android: [104, 0],
  safari: [16, 0],
  safari_ios: [16, 0],
} as Record<string, [number, number]>;

export function version(browser: [number, number]): number {
  return (browser[0] << 16) | (browser[1] << 8);
}
export function versionString(browser: [number, number]): string {
  return `${browser[0]}.${browser[1]}`;
}
