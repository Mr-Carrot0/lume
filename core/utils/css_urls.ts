type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    const quotedUrl = url.startsWith("data:")
      ? `'${newUrl.replaceAll("'", "\\'")}'`
      : `"${newUrl.replaceAll('"', '\\"')}"`;
    result += `${code.slice(index, start)}${quotedUrl}`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    let url = code.slice(start).trim();
    const quote = url.startsWith('"') || url.startsWith("'")
      ? url[0]
      : undefined;
    url = quote ? url.slice(1) : url;
    let end = findUrlEnd(url, quote);
    url = url.slice(0, end);
    end += start + (quote ? 2 : 0);
    url = url.replace(/\\'/g, "'").replace(/\\"/g, '"');
    return ["url", url, start, end];
  }
}

function findUrlEnd(code: string, quote?: string): number {
  if (!quote) {
    return code.indexOf(")");
  }

  let end = code.indexOf(quote);

  while (end !== -1 && code[end - 1] === "\\") {
    end = code.indexOf(quote, end + 1);
  }

  return end === -1 ? code.indexOf(")") : end;
}
