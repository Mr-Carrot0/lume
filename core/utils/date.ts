/**
 * Returns the date of the git commit that created or modified the file.
 * Thanks to https://github.com/11ty/eleventy/blob/8dd2a1012de92c5ee1eab7c37e6bf1b36183927e/src/Util/DateGitLastUpdated.js
 */
export function getGitDate(
  type: "created" | "modified",
  file: string,
): Date | undefined {
  const args = type === "created"
    ? ["log", "--diff-filter=A", "--follow", "-1", "--format=%at", "--", file]
    : ["log", "-1", "--format=%at", "--", file];

  const { stdout, success } = new Deno.Command("git", { args }).outputSync();

  if (!success) {
    return;
  }
  const str = new TextDecoder().decode(stdout);

  if (str) {
    return parseDate(parseInt(str) * 1000);
  }
}

/** Parse a string or number (of miliseconds) to UTC Date */
export function parseDate(date: string | number): Date {
  return new Date(getZonedDateTime(date).epochMilliseconds);
}

/** Parse a string or number (of miliseconds) to a zoned datetime */
export function getZonedDateTime(
  date: string | number,
  timezone = "UTC",
): Temporal.ZonedDateTime {
  if (typeof date === "number") {
    return Temporal.Instant.fromEpochMilliseconds(date).toZonedDateTimeISO(
      timezone,
    );
  }

  try {
    return Temporal.Instant.from(date).toZonedDateTimeISO(timezone);
  } catch {
    return Temporal.PlainDateTime.from(date).toZonedDateTime(timezone);
  }
}
