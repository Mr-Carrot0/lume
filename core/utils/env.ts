const envVars = new Map<string, string>();

export function setEnv(name: string, value: string) {
  if (allowedEnvVars()) {
    Deno.env.set(name, value);
  }
  envVars.set(name, value);
}

export function env<T>(name: string): T | undefined {
  const value = allowedEnvVars()
    ? envVars.get(name) ?? Deno.env.get(name)
    : envVars.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}

let allowed: boolean | undefined;
function allowedEnvVars(): boolean {
  if (allowed === undefined) {
    allowed = Deno.permissions.querySync?.({ name: "env" }).state === "granted";
  }
  return allowed;
}
