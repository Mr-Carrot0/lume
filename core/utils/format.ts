const byteUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "byte",
  unitDisplay: "narrow",
});

export function bytes(bytes: number) {
  return byteUnit.format(bytes);
}

const percentageUnit = Intl.NumberFormat("en", {
  style: "percent",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

export function percentage(initialValue: number, finalValue: number) {
  if (initialValue === finalValue) {
    return "0%";
  }

  const ratio = -(1 - finalValue / initialValue);

  if (ratio < 0) {
    return percentageUnit.format(ratio);
  }

  return "+" + percentageUnit.format(ratio);
}

const secondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "second",
  unitDisplay: "short",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

const milisecondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "millisecond",
  unitDisplay: "short",
  maximumFractionDigits: 0,
});

export function duration(miliseconds: number) {
  if (miliseconds < 1000) {
    return milisecondUnit.format(miliseconds);
  }

  return secondUnit.format(miliseconds / 1000);
}
