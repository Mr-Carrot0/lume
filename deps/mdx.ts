export * from "npm:@mdx-js/mdx@3.1.1";
export { default as remarkGfm } from "npm:remark-gfm@4.0.1";
