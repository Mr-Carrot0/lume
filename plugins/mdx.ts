import loader from "../core/loaders/text.ts";
import { merge } from "../core/utils/object.ts";
import { compile, remarkGfm } from "../deps/mdx.ts";
import { join, toFileUrl } from "../deps/path.ts";
import { renderComponent } from "../deps/ssx.ts";

import type Site from "../core/site.ts";
import type { Engine } from "../core/renderer.ts";
import type { PluggableList, RehypeOptions } from "../deps/remark.ts";

export interface Options {
  /** File extensions to load */
  extensions?: string[];

  /** List of recma plugins to use */
  recmaPlugins?: PluggableList;

  /**
   * List of remark plugins to use
   * @default `[remarkGfm]`
   */
  remarkPlugins?: PluggableList;

  /** Options to pass to rehype */
  rehypeOptions?: RehypeOptions;

  /** List of rehype plugins to use */
  rehypePlugins?: PluggableList;

  /** Set `false` to remove the default plugins */
  useDefaultPlugins?: boolean;

  /** Components to add/override */
  components?: Record<string, unknown>;

  /**
   * Custom includes path
   * @default `site.options.includes`
   */
  includes?: string;
}

// Default options
export const defaults: Options = {
  extensions: [".mdx"],
  useDefaultPlugins: true,
};

const remarkDefaultPlugins = [
  remarkGfm,
];

/** Template engine to render Markdown files with Remark */
export class MDXEngine implements Engine<string | { toString(): string }> {
  baseUrl: string;
  options: Required<Options>;
  includes: string;

  constructor(baseUrl: string, options: Required<Options>) {
    this.baseUrl = baseUrl;
    this.options = options;
    this.includes = options.includes;
  }

  deleteCache() {}

  async render(
    content: string,
    data?: Record<string, unknown>,
    filename?: string,
  ) {
    const baseUrl = toFileUrl(join(this.baseUrl, filename || "/")).href;
    const pragma = `/** @jsxImportSource lume */`;
    const result = await compile(content, {
      baseUrl,
      jsx: true,
      format: "mdx",
      outputFormat: "function-body",
      recmaPlugins: this.options.recmaPlugins,
      remarkPlugins: this.options.remarkPlugins,
      rehypePlugins: this.options.rehypePlugins,
      remarkRehypeOptions: this.options.rehypeOptions,
      stylePropertyNameCase: "css",
    });

    const destructure = `{${Object.keys(data!).join(",")}}`;
    const code = result.toString()
      .replace("/*@jsxRuntime automatic*/\n/*@jsxImportSource react*/", pragma)
      .replace(
        '"use strict";\n',
        `export default async function (${destructure}) {`,
      ) +
      "}";

    const url = URL.createObjectURL(new Blob([code], { type: "text/jsx" }));
    const module = (await import(url)).default;
    const mdxContext = (await module(data)).default;
    URL.revokeObjectURL(url);

    const body = mdxContext({
      components: { comp: data?.comp, ...this.options.components },
    });

    return renderComponent(body);
  }

  addHelper() {}
}

/**
 * A plugin to render MDX files
 * @see https://lume.land/plugins/mdx/
 */
export function mdx(userOptions?: Options) {
  return function (site: Site) {
    const options = merge(
      { ...defaults, includes: site.options.includes },
      userOptions,
    );

    if (options.useDefaultPlugins) {
      options.remarkPlugins ||= [];
      options.remarkPlugins.unshift(...remarkDefaultPlugins);
    }

    const engine = new MDXEngine(site.src(), options);

    // Ignore includes folder
    if (options.includes) {
      site.ignore(options.includes);
    }

    // Load the pages and register the engine
    site.loadPages(options.extensions, {
      loader,
      engine,
    });

    // Register the filter
    const filter = async (
      content: string,
      data?: Record<string, unknown>,
    ): Promise<string> =>
      (await engine.render(content, data)).toString().trim();

    site.filter("mdx", filter, true);
  };
}

export default mdx;

/** Extends Helpers interface */
declare global {
  namespace Lume {
    export interface Helpers {
      /** @see https://lume.land/plugins/mdx/ */
      mdx: (content: string, data?: Record<string, unknown>) => Promise<string>;
    }
  }
}
