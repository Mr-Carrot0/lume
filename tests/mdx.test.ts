import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import mdx from "../plugins/mdx.ts";
import jsx from "../plugins/jsx.ts";

Deno.test("Build a mdx site", async (t) => {
  const site = getSite({
    src: "mdx",
  });

  const plugin = () => {
    // deno-lint-ignore no-explicit-any
    return (tree: any) => {
      tree.children.push({
        type: "element",
        tagName: "div",
        properties: { style: "background-color: orange;" },
        children: [{ type: "text", value: "Hello, World!" }],
      });
    };
  };

  site.use(jsx());
  site.use(mdx({ rehypePlugins: [plugin] }));

  await build(site);
  await assertSiteSnapshot(t, site);
});
