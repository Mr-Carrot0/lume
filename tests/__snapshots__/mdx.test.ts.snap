export const snapshot = {};

snapshot[`Build a mdx site 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      engines: 1,
      ext: ".jsx",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tsx",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".mdx",
      isPage: true,
      loader: [AsyncFunction: text],
    },
  ],
  src: [
    "/",
    "/_components",
    "/_components/Header.jsx",
    "/_includes",
    "/_includes/Image.tsx",
    "/index.mdx",
    "/mdx-filter.page.ts",
  ],
}
`;

snapshot[`Build a mdx site 2`] = `[]`;

snapshot[`Build a mdx site 3`] = `
[
  {
    content: '<header><h1>Hello world</h1><p>This is a description</p></header>
<h2>Hello world</h2>
<p>This is a markdown file with the title <strong>Hello world</strong>.</p>
<img src="https://via.placeholder.com/350x150" alt="foo">',
    data: {
      basename: "index",
      children: '<header><h1>Hello world</h1><p>This is a description</p></header>
<h2>Hello world</h2>
<p>This is a markdown file with the title <strong>Hello world</strong>.</p>
<img src="https://via.placeholder.com/350x150" alt="foo">',
      comp: [
        "_components",
        "_proxies",
      ],
      content: 'import Image from "./_includes/Image.tsx";

<comp.Header title={title} description={description}/>

## Hello world

This is a markdown file with the title **{ title }**.

<Image alt="foo" />
',
      date: [],
      description: "This is a description",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Hello world",
      url: "/",
    },
    src: {
      ext: ".mdx",
      path: "/index",
      remote: undefined,
    },
  },
  {
    content: '<hr>
<h2>title: Hello world
description: This is a description</h2>

<header><h1>Hello world</h1><p>This is a description</p></header>
<h2>Hello world</h2>
<p>This is a markdown file with the title <strong>Hello world</strong>.</p>
<img src="https://via.placeholder.com/350x150" alt="foo">',
    data: {
      basename: "mdx-filter",
      children: '<hr>
<h2>title: Hello world
description: This is a description</h2>

<header><h1>Hello world</h1><p>This is a description</p></header>
<h2>Hello world</h2>
<p>This is a markdown file with the title <strong>Hello world</strong>.</p>
<img src="https://via.placeholder.com/350x150" alt="foo">',
      comp: [
        "_components",
        "_proxies",
      ],
      content: "default",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "mdx filter example",
      url: "/mdx-filter/",
    },
    src: {
      ext: ".page.ts",
      path: "/mdx-filter",
      remote: undefined,
    },
  },
]
`;
